### ISSUE TITLE
WSD loss doesn't decrease

### ISSUE THREAD

**Issue by ztwe** [Yochengliu/MLIC-KD-WSD#8]
solver:
```
net: "wsddn/wsddn_train.prototxt"
test_iter: 800
test_interval: 2400
base_lr: 0.0001
momentum: 0.9
type: "SGD"
weight_decay: 0.0005
lr_policy: "inv"
gamma: 0.00001
power: 0.75
display: 200
max_iter: 800000
snapshot: 50000
snapshot_prefix: "models"
solver_mode: GPU
```
train batch_size=3

When iteration comes to 30000, Loss varies between 20 and 50.
Have you encounter the problem?

**Comment by Yochengliu** [comment]
No. Did you finetune with ImageNet pretrained model? If so, you can try to decrease the learning rate. We use base_lr = 0.00005 to train WSDDN.

**Comment by ztwe** [comment]
Yes, I finetune with the pretrian model.

 The difference between 0.00005 and 0.0001 seems really tiny.
The scores of all 500 proposals are 1.
Does batch_size result in the problem?

I have tried moving your custom code to this multi-gpu caffe project.
https://github.com/sciencefans/CaffeMex_v2

But it still works in single GPU.

Now it comes to 90000 iteration which means the third epoch.
The minimum of loss is 12. The maximum is 70+.
But it still mainly varies between 20 and 50. 

WSDDN seems really hard to reproduce.

**Comment by Yochengliu** [comment]
I think it may not be caused by the batch_size. If all the scores are 1, you can set score scale as 1 and score bias as 0 in prototxt. Please make sure you have modified the prototxt according to your dataset (classes: 80->yours, boxes number: 500->yours), and you have checked the datalist, finetuning process, carefully.  You can set a small lr_rate as 1e-5. As a reference for you, the initial loss for WSDDN in our experiments is always less than 10, and it converges quickly even with batch_size = 1. 

I also have no experience on multi-gpu Caffe, sorry to have no idea for this issue.

**Comment by ztwe** [comment]
![image](https://user-images.githubusercontent.com/7899459/49623336-31622c80-fa08-11e8-9804-2bda6d3e9331.png)

Now the lr_rate is setted to 1e-5.

The initial loss is about 10+ or 20+ which means the loading of weights from pretrain model fails.
But the info in screenshot shows that conversion between old caffemodel and new caffe is successful.

Classes and boxes number are setted as following:
```
    att_num: 80
    bbox_num: 500
```

**Comment by Yochengliu** [comment]
Please show your datalist with a screenshot, solver and prototxt, and tell me how you preprocess the images and how you acquire the bounding boxes. I will try my best to find the issue.

**Comment by ztwe** [comment]
You are really nice. I'm sorry to disturb you.
1. datalist
![image](https://user-images.githubusercontent.com/7899459/49624913-5bb6e880-fa0e-11e8-9953-faa6f980ac53.png)

![image](https://user-images.githubusercontent.com/7899459/49624972-9751b280-fa0e-11e8-9261-1c4ea5c65bc1.png)


2. solver:
```
net: "wsddn/wsddn_train.prototxt"
test_iter: 800
test_interval: 2400
base_lr: 0.00001
momentum: 0.9
type: "SGD"
weight_decay: 0.0005
lr_policy: "inv"
gamma: 0.00001
power: 0.75
display: 200
max_iter: 800000
snapshot: 50000
snapshot_prefix: "models"
solver_mode: GPU
```


3. prototxt
```
name: "VGG16_wsddn"

layer{
  name: "data"
  type: "HumanAttData"
  top: "data"
  top: "label_roi"
  include {
    phase: TRAIN
  }
  transform_param {
    mirror: true
    mean_value: 103.52
    mean_value: 116.28
    mean_value: 123.675
  }
  human_att_data_param {
    source: "train_672_mscoco_withebscore.txt"
    root_folder: ""
    batch_size: 3
    shuffle: true
    new_width: 672
    new_height: 672
    scale_factors: 0.714286  # 480  30
    scale_factors: 0.857143  # 576  36
    scale_factors: 1         # 672  42
    scale_factors: 1.095239  # 736  46
    scale_factors: 1.238096  # 832  52
    att_num: 80
    bbox_num: 500
  }
}
```


4. the code of acquire boxes
```
frame = cv2.imread(file_name)
frame = frame.astype(np.float32) 
frame = np.divide(frame, 255.0)

edgeGenerator = cv2.ximgproc.createStructuredEdgeDetection(model = modelfile)
edgearray = edgeGenerator.detectEdges(frame )
orientationarray = edgeGenerator.computeOrientation(edgearray)
suppressed_edgearray = edgeGenerator.edgesNms(edgearray, orientationarray)


boxGenerator = cv2.ximgproc.createEdgeBoxes(maxBoxes = 500,
                                                    alpha = 0.65,
                                                    beta = 0.75,
                                                    minScore = 0.03)
boxes = boxgenerator.getBoundingBoxes(suppressed_edgearray, orientationarray)

if(len(boxes) < 500):
    if len(boxes) > 250:
      boxes = complete_boxes_list(boxes)
    else:
      return boxes

  boxes_array = np.array(boxes)

  boxes_array[:, 2] = boxes_array[:, 2] + boxes_array[:, 0] # x2
  boxes_array[:, 3] = boxes_array[:, 3] + boxes_array[:, 1] # y2

  boxes_insert = np.insert(boxes_array, [4], [1], axis=1) # score

  boxes_reshape = boxes_insert.reshape((-1))
```

**Comment by Yochengliu** [comment]
It seems no issues. Please set the score_scale as 1 and score_bias as 0, and try smaller lr_rate like 1e-6 and 1e-7.

**Comment by ztwe** [comment]
Where and wow to set score_scale  and score_bias?

**Comment by Yochengliu** [comment]
Search in the prototxt, you will find them.

**Comment by ztwe** [comment]
![image](https://user-images.githubusercontent.com/7899459/49635907-14475100-fa3c-11e8-9bfc-65e1f4afb67a.png)

**Comment by ztwe** [comment]
The training is blocked at iter 23000 for 2 hours. It has been stopped by ctrl+c.
What may cause this problem? image file? Or others?

**Comment by Yochengliu** [comment]
Our best test loss is 3.98. We train WSDDN for 120 epochs. 

If you didn't use the Caffe-MLIC, it may be caused by the Caffe you used. Otherwise, it may be caused by image file, gpu or server. You can training without shuffle data, and check whether the bug would be reproduced stably at a fixed iteration. If so, you could locate the image file according to this iteration.

**Comment by ztwe** [comment]
Now iteration comes to 110000 which means the 4th epoch, and the loss of test varies between 5.5 and 6.
Is that right? If not, how to decrease?

The reason of being blocked may be the insufficient of PC memory.

**Comment by Yochengliu** [comment]
Sorry, we can not judge whether your test loss is right, because the setting you use is not consistent with ours. We can only suggest you train for more epochs, and possibly, try other lr_rate strategies like multi-step.

**Comment by ztwe** [comment]
Thx!
